src/samples/iio+network/iio-gyroscope-console-and-mqtt-publish.c

```diff
@@ -0,0 +1,542 @@
+        pivot_index = partition(temp, left, right, pivot_index);
+        if (pivot_index == median_index)
+            return temp[median_index];
+        else if (pivot_index > median_index)
+            right = pivot_index - 1;
+        else
+            left = pivot_index + 1;
+    }
+
+    return temp[left];
+}
+
+static void
+denoise_median(struct filter_median_t *filter_data, double *x, double *y, double *z)
+{
+    /* Thanks to https://github.com/01org/android-iio-sensors-hal
+     * for denoise algorithm
+     */
+    unsigned int offset;
+
+    if (filter_data->count < filter_data->sample_size)
+        filter_data->count++;
+
+    offset = 0;
+    filter_data->buff[offset + filter_data->idx] = *x;
+    *x = median(filter_data->buff + offset, filter_data->count);
+
+    offset = filter_data->sample_size * 1;
+    filter_data->buff[offset + filter_data->idx] = *y;
+    *y = median(filter_data->buff + offset, filter_data->count);
+
+    offset = filter_data->sample_size * 2;
+    filter_data->buff[offset + filter_data->idx] = *z;
+    *z = median(filter_data->buff + offset, filter_data->count);
+
+    filter_data->idx = (filter_data->idx + 1) % filter_data->sample_size;
+}
+
+#endif
+
+static void
+iio_gyroscope_reader_cb(void *data, struct sol_iio_device *device)
+{
+    int r;
+    char buffer[128];
+    struct iio_gyroscope_data *gyro_data = data;
+    struct sol_direction_vector out;
+
+#ifdef MQTT
+    struct sol_buffer payload;
+    struct sol_mqtt_message mqtt_message;
+#endif
+#ifdef GYRO_CALIBRATE
+    struct gyro_cal_t *cal_data = &(gyro_data->cal_data);
+#endif
+#ifdef DENOISE_MEDIAN
+    struct filter_median_t *filter_data = &(gyro_data->filter_data);
+#endif
+    r = sol_iio_read_channel_value(gyro_data->channel_x, &out.x);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    r = sol_iio_read_channel_value(gyro_data->channel_y, &out.y);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    r = sol_iio_read_channel_value(gyro_data->channel_z, &out.z);
+    SOL_INT_CHECK_GOTO(r, < 0, error);
+
+    /*
+     * For noisy sensors drop a few samples to make sure we have at least
+     * GYRO_DROP_SAMPLES events in the filtering queue.
+     * This improves mean and std dev.
+     */
+    if (gyro_data->drop_samples_count < GYRO_DROP_SAMPLES) {
+        gyro_data->drop_samples_count++;
+        return;
+    }
+
+    sol_iio_mount_calibration(device, &out);
+
+#ifdef GYRO_CALIBRATE
+    if (cal_data->calibrated == false)
+        cal_data->calibrated = gyro_collect(cal_data, out.x, out.y, out.z);
+
+    out.x = out.x - cal_data->bias_x;
+    out.y = out.y - cal_data->bias_y;
+    out.z = out.z - cal_data->bias_z;
+#endif
+
+#ifdef DENOISE_MEDIAN
+    denoise_median(filter_data, &out.x, &out.y, &out.z);
+#endif
+
+#ifdef GYRO_CALIBRATE
+    clamp_gyro_readings_to_zero(cal_data, &out.x, &out.y, &out.z);
+#endif
+
+#ifdef GYRO_CALIBRATE
+    snprintf(buffer, 127, "%f\t%f\t%f\t[rad/sec]\t(%s)", out.x, out.y, out.z,
```

Comment:
Maybe use `sizeof(buffer) -1`, so in case buffer size changes, code is still correct.